### PR TITLE
Fix server crash when a client disappears shortly after login

### PR DESCRIPTION
**Pull request by zeuner**
There seems to be a race condition when a client logs in, and is closed while the resources are loaded:

2020-04-21 18:19:01: ERROR[Main]: ServerError: AsyncErr: ServerThread::run Lua: Runtime error from mod 'default' in callback on_joinplayer(): ...ine/bin/../games/minetest-game/mods/default/init.lua:23: attempt to index local 'info' (a nil value)
2020-04-21 18:19:01: ERROR[Main]: stack traceback:
2020-04-21 18:19:01: ERROR[Main]:       ...ine/bin/../games/minetest-game/mods/default/init.lua:23: in function '?'
2020-04-21 18:19:01: ERROR[Main]:       ...ock/minetest-engine/bin/../builtin/game/register.lua:429: in function <...ock/minetest-engine/bin/../builtin/game/register.lua:413>

This patch should avoid the crash.

Maybe it makes sense to check whether the engine can be prevented from calling the handler in these cases altogether.

### CODE REVIEW

**Comment by sfan5** [comment]
Relevant engine issue: https://github.com/minetest/minetest/issues/9352

**Comment by paramat** [comment]
I do not think it can be assumed and reported that the player has 'disappeared', are you assuming that because 'info' is 'nil'? We have had no reports of players 'disappearing' related to this.
See the engine issue, the cause of 'info' being 'nil' might be something else. Previously, before the recent attempt at a fix, the cause was RTT values.

If you 'return' at this point in code, then a crash is avoided, but then neither variation of formspec prepend is set, breaking formspecs, and hotbar textures are not set, which are a problem if the player actually still exists.
Also, see the comments in #2586 i doubt your PR is acceptable for those reasons.

:-1:

**Comment by sfan5** [comment]
There is nothing wrong with the code if the player <i>actually</i> disappears, but nobody has debugged that yet for the new (post 5.2.0) variant of this bug.

File: mods/default/init.lua [hunk 1]
```diff
@@ -20,6 +20,10 @@ minetest.register_on_joinplayer(function(player)
 			listcolors[#00000069;#5A5A5A;#141318;#30434C;#FFF] ]]
 	local name = player:get_player_name()
 	local info = minetest.get_player_information(name)
+	if not info then
+		minetest.log("warning", "player " .. name .. " disappeared")
+		return
+	end
 	if info.formspec_version > 1 then
```
Comment (MoNTE48) [inline mods/default/init.lua:27]: Just 
`if info and info.formspec_version > 1 then`

**Comment by zeuner** [comment]
Thanks for the pointer!

I agree there may be better places to fix this.

However, one might argue that a mod/subgame developer _should_ currently check whether minetest.get_player_information returns nil since the API docs don't specify whether it is guaranteed to be non-nil within a handler registered through minetest.register_on_joinplayer.

It takes quite some attempts to reproduce it, but here you go with a relevant log excerpt:

2020-04-24 22:55:16: ACTION[Server]: test23 joins game. List of players: test23
2020-04-24 22:55:16: [Server]: static int ModApiServer::l_get_player_information(lua_State*): peer was not found
2020-04-24 22:55:16: WARNING[Server]: player test23 disappeared
2020-04-24 22:55:17: ACTION[Server]: test23 leaves game. List of players:

**Comment by zeuner** [comment]
@paramat: Good point. I assumed it was because the player left because the bug only occurred for me in these cases, but this does not imply that it cannot happen in other cases.

However, you might find it interesting that on the server where I found this, minetest.get_player_information returned nil two times out of 32 logins where the player immediately left, while there were 276 logins altogether. So far, it never happened when the player did not immediately leave. I know, correlation does not mean causation, but maybe it serves as a clue for something?

File: mods/default/init.lua (hunk 1 above)
Comment (zeuner) [reply mods/default/init.lua:27]: This would hide that something fishy is going on. I have to figure out whether returning from the handler is the best thing to do, though.

**Comment by paramat** [comment]
Ok thanks, it does seem likely to be when a player disappears then. The engine issue suggests this happens too.

File: mods/default/init.lua (hunk 1 above)
Comment (paramat) [reply mods/default/init.lua:27]: MoNTE488, my PR #2586 did exactly that and was disapproved.

**Comment by sfan5** [comment]
@zeuner thanks for confirming my suspicion.
Since you have the ability to reproduce the issue, would you mind testing minetest/minetest#9739?

**Comment by zeuner** [comment]
Another test run from today: 389 logins, average time between join and leave: 24 seconds, 3 times info was nil, with 0-1 seconds between join and leave in these 3 cases.

@sfan5: I will test again using sfan5:playerinfo_fix in the engine and an unpatched minetest_game. But I assume this will work fine if info will now always have a formspec_version entry.

**Comment by zeuner** [comment]
@sfan5: As far as I can see, this one can be closed if https://github.com/minetest/minetest/pull/9739 gets merged. Using minetest_game:master and the patch from https://github.com/minetest/minetest/pull/9739, I neither get the crash due to info being nil, nor the "peer was not found" warning. This being observed after 128 connections with 0 seconds between join and leave.

**Comment by sfan5** [comment]
Thanks for testing and confirming that the fix works.
